document stage:add (#2208)
iterative/dvc.org: content/docs/command-reference/stage/add.md

> # run

Reviewer: title?

> Helper command to add or update <abbr>stages</abbr> in `dvc.yaml`. Requires a
> name and a command.

Reviewer 2: ```suggestion
Helper command to create or update <abbr>stages</abbr> in `dvc.yaml`.
```

Author: I'd prefer `add` as that's what the command name is.

## Synopsis

```usage
usage: dvc stage add [-h] [-q | -v] -n <name> [-d <path>] [-o <path>]

>                [-O <path>] [-p [<path>:]<params_list>] [-m <path>]

Reviewer: indent?

                     [-M <path>] [--plots <path>] [--plots-no-cache <path>]
                     [-w <path>] [-f]
                     [--outs-persist <path>] [--outs-persist-no-cache <path>]
                     [--always-changed] [--external] [--desc <text>]

>                     command

Author: If I keep the `command` indented, the syntax highlighting fails. If I add one space or remove it, it mysteriously works.

Reviewer: it's probably a bug as well - mind creating a ticket? with this one and the previous.

Reviewer 2: I tried this locally (added a space before `command` to align it) and the commit checks passed. Can you clarify please? 🙂

Reviewer 2: Ah I see you were talking about syntax highlighting. Nvmd

> usage: dvc stage add [-h] [-q | -v] -n <name> [-d <path>] [-o <path>]
>                      [-O <path>] [-p [<path>:]<params_list>] [-m <path>]
>                      [-M <path>] [--plots <path>] [--plots-no-cache <path>]
>                      [-w <path>] [-f]
>                      [--outs-persist <path>] [--outs-persist-no-cache <path>]
>                      [--always-changed] [--external] [--desc <text>]
>                     command

Reviewer 2: > syntax highlighting fails. If I add one space or remove it, it mysteriously works

The lines are too long anyway, please try:
```suggestion
usage: dvc stage add [-h] [-q | -v] -n <name> [-d <path>] [-o <path>]
                     [-O <path>] [-p [<path>:]<params_list>]
                     [-m <path>] [-M <path>] [--plots <path>]
                     [--plots-no-cache <path>] [-w <path>] [-f]
                     [--outs-persist <path>]
                     [--outs-persist-no-cache <path>]
                     [--always-changed] [--external] [--desc <text>]
                     command
```

Author: Yup, that works. Thanks.

positional arguments:

>   command               Command to execute.

Reviewer 2: ```suggestion
  command               Command to execute
```

```

## Description

> `dvc stage add` is a helper for creating or updating
> [pipeline](/doc/command-reference/dag) stages in a `dvc.yaml` file (located in
> the current working directory).

Reviewer 2: ```suggestion
Creates or updates stages in a [pipeline](/doc/command-reference/dag) (saved to
`dvc.yaml` in the current working directory).
```

Reviewer 2: But how do you update vs. create?

Author: `update` just means overwrite here (with `--force`).

Reviewer 2: Sure. My question was meant as a hint on what to improve in the explanation.

Reviewer 2: Which at this point (since this is merge) I'll do separately, so np.

_Stages_ represent individual data processes, including their input and
resulting outputs. They can be combined to capture simple data workflows,

> organize data science projects, or build detailed machine learning pipelines.

Reviewer 2: Not sure we want to repeat this here again since it's already in the index and in the tooltip.

A stage name is required and can be provided using the `-n` (`--name`) option.

> The other available [options](#options) are mostly meant to describe different
> kinds of stage [dependencies and outputs](#dependencies-and-outputs). The
> remaining terminal input provided to `dvc stage add` after `-`/`--` flags will
> become the required [`command` argument](#the-command-argument).

Reviewer 2: Copy edit (committing):
```suggestion
Most of the other [options](#options) help with defining different kinds of
[dependencies and outputs](#dependencies-and-outputs) for the stage. The
remaining terminal input provided to `dvc stage add` after `-`/`--` flags will
become the required [`command` argument](#the-command-argument).
```

> `dvc stage add`, unlike `dvc run`, does not execute stage commands by default.
> `dvc repro` command can be used instead to execute them.
> 

Reviewer 2: If we're gonna remove the `run` ref in this PR as well (I think Ivan suggested that), maybe just say:
```suggestion
`dvc repro` can be used to execute pipelines after their stages have been defined.
```
And since it's such a short sentence, probably put it into some existing paragraph before.

Reviewer 2: > And since it's such a short sentence, probably put it into some existing paragraph before.

I'll do this part.

  [add missing dependencies/outputs to an existing stage](/docs/user-guide/how-to/add-deps-or-outs-to-a-stage)
  without having to execute it again.

- Renaming dependencies or outputs requires a
  [manual process](/doc/command-reference/move#renaming-stage-outputs) to update
  `dvc.yaml` and the project's cache accordingly.

### For displaying and comparing data science experiments

[parameters](/doc/command-reference/params) (`-p`/`--params` option) are a
special type of key/value dependencies. Multiple parameter dependencies can be
specified from within one or more YAML, JSON, TOML, or Python parameters files
(e.g. `params.yaml`). This allows tracking experimental hyperparameters easily.

Special types of output files, [metrics](/doc/command-reference/metrics) (`-m`
and `-M` options) and [plots](/doc/command-reference/plots) (`--plots` and
`--plots-no-cache` options), are also supported. Metrics and plots files have
specific formats (JSON, YAML, CSV, or TSV) and allow displaying and comparing
data science experiments.

### The `command` argument

The `command` sent to `dvc stage add` can be anything your terminal would accept

> and run directly, for example a shell built-in, expression, or binary found in

Reviewer 2: I know this comes exactly from the `run` ref. but should we take the opportunity to move this before `### Dependencies and outputs`?

`PATH`. Please remember that any flags sent after the `command` are considered
part of the command itself, not of `dvc stage add`.

⚠️ While DVC is platform-agnostic, the commands defined in your
[pipeline](/doc/command-reference/dag) stages may only work on some operating
systems and require certain software packages to be installed.

Wrap the command with double quotes `"` if there are special characters in it
like `|` (pipe) or `<`, `>` (redirection), otherwise they would apply to
`dvc stage add` itself. Use single quotes `'` instead if there are environment
variables in it that should be evaluated dynamically. Examples:

```dvc
$ dvc stage add -n first_stage "./a_script.sh > /dev/null 2>&1"
$ dvc stage add -n second_stage './another_script.sh $MYENVVAR'
```

## Options

- `-n <stage>`, `--name <stage>` (**required**) - specify a name for the stage

>   generated by this command (e.g. `-n train`). Stage names can only contain

Reviewer 2: Quick Q: Are there any changes to options text vs. https://dvc.org/doc/command-reference/run#options ?

Author: No changes on options text, just few options that were related to running  a stage are not in stage:add which were removed.

  `command`.

- `-f`, `--force` - overwrite an existing stage in `dvc.yaml` file without
  asking for confirmation.

- `--always-changed` - always consider this stage as changed (uses the
  `always_changed` field in `dvc.yaml`). As a result `dvc status` will report it
  as `always changed` and `dvc repro` will always execute it.

  > Note that regular `.dvc` files (without dependencies) are automatically
  > considered "always changed", so this option has no effect in those cases.

- `--external` - allow writing outputs outside of the DVC repository. See
  [Managing External Data](/doc/user-guide/managing-external-data).

- `--desc <text>` - user description of the stage (optional). This doesn't  
  affect any DVC operations.

- `-h`, `--help` - prints the usage/help message, and exit.

- `-q`, `--quiet` - do not write anything to standard output. Exit with 0 if no
  problems arise, otherwise 1.

- `-v`, `--verbose` - displays detailed tracing information.

## Examples

Let's create a <abbr>DVC project</abbr> and a stage (that counts the number of
lines in a `test.txt` file):

```dvc
$ mkdir example && cd example
$ git init
$ dvc init
$ mkdir data

> $ dvc stage add -n count \
>           -d test.txt \
>           -o lines \
>           "cat test.txt | wc -l > lines"

Reviewer 2: I assume there are no changes to the examples so won't review deeply. But 💅 can we please re-indent the code samples? E.g.:
```suggestion
$ dvc stage add -n count \
                -d test.txt \
                -o lines \
                "cat test.txt | wc -l > lines"
```
